test/powershell/Language/Parser/Conversions.Tests.ps1

```diff
@@ -396,4 +396,89 @@ Describe 'method conversion' -Tags 'CI' {
+        [string] GetB([datetime] $d)                                   { return "GetB-datetime-string" }
+
+        [string] GetBPrime([System.IO.FileSystemInfo] $fsinfo, [object] $o) { return "GetBPrime-filesysteminfo-object-string" }
+        [object] GetBPrime([System.IO.FileInfo] $finfo, [string] $s)        { return "GetBPrime-fileinfo-string-object" }
+        [string] GetBPrime([datetime] $d)                                   { return "GetBPrime-datetime-string" }
+
+        static [string] GetBStatic([System.IO.FileSystemInfo] $fsinfo, [object] $o) { return "GetBStatic-filesysteminfo-object-string" }
+        static [object] GetBStatic([System.IO.FileInfo] $finfo, [string] $s)        { return "GetBStatic-fileinfo-string-object" }
+        static [string] GetBStatic([datetime] $d)                                   { return "GetBStatic-datetime-string" }
+
+        ## Test enum parameter type
+        [object] GetC([E] $e) { return $e.ToString() }
+    }
+
+    It "Different method overloads with same signatures/orders should have same PSMethod type" {
+        $n = [N]::new()
+
+        $n.GetA.GetType() | Should -Be ($n.GetAPrime.GetType())
+        $n.GetA.GetType() | Should -Be ([N]::GetAStatic.GetType())
+
+        $n.GetB.GetType() | Should -Be ($n.GetBPrime.GetType())
+        $n.GetB.GetType() | Should -Be ([N]::GetBStatic.GetType())
+    }
+
+    It "Match signature with variance and use the first match when there is no exact match" {
+        $n = [N]::new()
+
+        [Func[[System.IO.FileInfo], [string], [object]]] $f = $n.GetA
+        $f.Invoke($null, $null) | Should -BeExactly "GetA-filesysteminfo-object-string"
+
+        $f = $n.GetAPrime ## $n.GetAPrime has the same type as $n.GetA, so it should hit the conversion cache
+        $f.Invoke([System.IO.FileInfo]::new("aaa"), "bbb") | Should -BeExactly "GetAPrime-filesysteminfo-object-string"
+
+        $f = [N]::GetAStatic ## [N]::GetAStatic has the same type as $n.GetA, so it should hit the conversion cache
+        $f.Invoke($null, "") | Should -BeExactly "GetAStatic-filesysteminfo-object-string"
+    }
+
+    It "Exact match is preferred over match with variance" {
+        $n = [N]::new()
+
+        [Func[[System.IO.FileInfo], [string], [object]]] $f = $n.GetB
+        $f.Invoke($null, $null) | Should -BeExactly "GetB-fileinfo-string-object"
+
+        $f = $n.GetBPrime ## $n.GetBPrime has the same type as $n.GetB, so it should hit the conversion cache
+        $f.Invoke([System.IO.FileInfo]::new("ccc"), "ddd") | Should -BeExactly "GetBPrime-fileinfo-string-object"
+
+        $f = [N]::GetBStatic ## [N]::GetBStatic has the same type as $n.GetB, so it should hit the conversion cache
+        $f.Invoke($null, "") | Should -BeExactly "GetBStatic-fileinfo-string-object"
+    }
+
+    It "Test enum type parameter" {
+        $n = [N]::new()
+
+        [Func[[E], [object]]] $f = $n.GetC
+        $f.Invoke([E]::Week) | Should -BeExactly "Week"
+    }
```

Comment:
We should also include the following scenarios in tests to ensure we know current behaviour and detect regressions:

* Type constraint applied to an argument
    - `[object]$var = "Hi"; $cls.TestMethod($var)`
    - `$cls.TestMethod([object]"Hi")`
    - `[Super]$sub = [Sub]::new(); $cls.TestMethodWithSuperAndSubOverloads($sub)`
    - `$sub = [Sub]::new(); $cls.TestMethodWithSuperAndSubOverloads([Super]$sub)`
    - `[string]$fileName = [System.IO.FileInfo]::new("myFile"); $cls.TestMethodWithFileInfoAndStringOverloads($fileName)`
    - `$file = [System.IO.FileInfo]::new("myFile"); $cls.TestMethodWithFileInfoAndStringOverloads([string]$file)`
* Sub-class which matches two or more method signatures, but none exactly
```powershell
class Animal { }
class Cat : Animal { }

# If the order of method definition is what affects the method resolution using the current
# algorithm, we should have another test with the same methods in a different order on a different
# class to demonstrate that
class TestClass
{
    [string] TestMethod([object] $obj) { return "Object" }
    [string] TestMethod([Animal] $animal) { return "Animal" }
}
```
* Interfaces
    - Interface as parameter type in signature, object implements interface
    - Overload of method for multiple interfaces, object implements both interfaces
* Implicit coercions
    - I can't remember if we allow coercions in methods, but if so we must test how coercion works when
      there are multiple overloads we could coerce to.
* Maybe some scenarios where the method binding should fail, so we cover the failure mode code paths
    - No overload that matches
    - Generic methods